VACMS-19512: Drupal UX for phone numbers

davidmpickett opened the pull request:
## Description

Relates to #19512 
Also fixing #16150 as a bonus

## Testing done
Manual in Tugboat

## Screenshots

Entity & state | Current UI | Updated UI
-- | -- | --
Staff profile (no phone) |![Screenshot 2024-10-25 at 12 53 13 PM](https://github.com/user-attachments/assets/1a2384a6-393c-44b7-8ecb-758d15daba82) |![Screenshot 2024-10-24 at 5 16 25 PM](https://github.com/user-attachments/assets/9f0a3e81-5f9c-41a2-aacc-f862991d50df)
Staff profile (with phone) |![Screenshot 2024-10-24 at 5 19 03 PM](https://github.com/user-attachments/assets/935b7c6c-05f1-460f-82e4-ce0e95004235) |![Screenshot 2024-10-24 at 5 31 56 PM](https://github.com/user-attachments/assets/2f0c1fc7-039e-4821-ab76-6a4390c9c347)
VAMC Billing & insurance (with phone) |![Screenshot 2024-10-25 at 12 15 41 PM](https://github.com/user-attachments/assets/4c02366c-5472-4fc4-bf8d-bcd94e9c4f7d) |![Screenshot 2024-10-24 at 5 52 26 PM](https://github.com/user-attachments/assets/067ed6a0-8096-40a7-9bc6-deb0dd5e0680)
VAMC Billing & insurance (no phone) |![Screenshot 2024-10-25 at 12 15 29 PM](https://github.com/user-attachments/assets/9c9bb788-991a-4756-bb4f-7b6f0cae62c7) |![379959166-f3f7b956-3fa8-448e-b145-58b646920e8f](https://github.com/user-attachments/assets/a865d2cb-27ae-4526-9048-8ccd18da63d6)
VAMC Facility mental health (with phone) |![Screenshot 2024-10-25 at 12 11 38 PM](https://github.com/user-attachments/assets/577ca106-2c74-4bd8-9605-87d86da1c4c0) |![Screenshot 2024-10-25 at 12 49 46 PM](https://github.com/user-attachments/assets/0f611ac9-b891-47aa-b73e-6021237ade57)
VAMC Facility mental health (no phone) |![Screenshot 2024-10-25 at 12 11 19 PM](https://github.com/user-attachments/assets/9e0b5a1f-6ca7-43f9-b9a2-ee2b0d50aa0f) |![Screenshot 2024-10-25 at 12 49 32 PM](https://github.com/user-attachments/assets/e9260097-3266-4c80-a134-1a6ef59bea23)
VAMC Facility mental health (FEATURE TOGGLE OFF) |![Screenshot 2024-10-25 at 2 24 58 PM](https://github.com/user-attachments/assets/12d62334-5502-4eec-969b-e79a0f5b8421) |![Screenshot 2024-10-25 at 2 25 52 PM](https://github.com/user-attachments/assets/a374b438-2928-4919-a28a-1fa964758ba7)


## QA steps

What needs to be checked to prove this works?
What needs to be checked to prove it didn't break any related things?
What variations of circumstances (users, actions, values) need to be checked?

### Once per day setup
1. [Log in](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/) as an admin
2. Go to [Feature Toggle page](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/admin/config/system/feature_toggle) 
   - [x] Check FEATURE_TELEPHONE_MIGRATION_V1
   - [x] Click Save
3. Run Telephone migration (optional)
   - [x] Go to [Dashboard](https://tugboat.vfs.va.gov/671ac5aa25a4ad7d76bce8f3)
   - [x] Run each command below separately in php terminal
   - [x] Total migration time should be 30-40 minutes.

```
time drush codit-batch-operations:run MigratePhoneFieldToParagraph
time drush codit-batch-operations:run MigrateVamcBillingInsurancePhoneFieldToParagraph
time drush codit-batch-operations:run MigrateVamcFacilityMentalHealthPhoneFieldToParagraph
```

### Check Staff Profile
1. [Log in to tugboat](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/) as a VAMC editor or Content Admin
2. Go to Content > Add Content > [Staff Profile](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/node/add/person_profile)
3. Save a staff profile with no phone number
   - [x] Fill in the required fields (First name, last name, section, related office)
   - [x] In the contact information box, notice that by default, no phone number has been added
   - [x] Add a revision message and save as draft
4. Save a staff profile with a phone number
   - [x] Edit the staff profile you just created
   - [x] In the contact information box click the Add Phone number button
   - [x] Before filling out any of the phone number sub-fields, add a revision message and attempt save as draft
   - [x] See validation error for not filling out the required phone number sub-field 
   - [x] Fill in the required fields
   - [x] Save as draft
5. Remove a phone number from a staff profile
   - [x] Edit the staff profile again
   - [x] In the contact information box click the Remove button
   - [x] Click confirm removal button (this is technically optional)
   - [x] Add a revision message and save as draft

### Check VAMC System Billing & Insurance
1. [Log in to tugboat](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/) as a Content Admin
2. Go to Content > Add Content > [VAMC System Billing and Insurance](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/node/add/vamc_system_billing_insurance)
3. Save a Billing & Insurance page with a phone number
   - [x] Notice that in the "For inquiries by phone about copay balance" section, a phone number is added by default
   - [x] Before filling out any of the phone number sub-fields, add a revision message and attempt save as draft
   - [x] See validation error for not filling out the required phone number sub-field 
   - [x] Fill in other required fields
   - [x] Save as draft
4. Remove a phone number from a Billing & Insurance page
   - [x] Edit the page you just created
   - [x] In the "For inquiries by phone about copay balance" section click the Remove button
   - [x] Click confirm removal button (this is technically optional)
   - [x] Add a revision message and save as draft
5. Add a phone number 
   - [x] Edit the page you just created
   - [x] In the "For inquiries by phone about copay balance" section click the Add Phone number button
   - [x] Confirm that the correct phone number fields populate
6. Verify change for [16150](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/16150) 
    - [x] Edit the page you just created
    - [x] Notice that there is no longer an empty "Cashier's office" accordion

### Check VAMC Facility - Mental Health Phone Number
1. [Log in to tugboat](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/) as a Content Admin
2. Go to Content > Add Content > [VAMC Facility](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/node/add/health_care_local_facility)
3. Save a VAMC facility page with a phone number
   - [x] Notice that in the "Mental health phone number" section, a phone number is added by default
   - [x] Before filling out any of the phone number sub-fields, add a revision message and attempt save as draft
   - [x] See validation error for not filling out the required phone number sub-field 
   - [x] Fill in other required fields
   - [x] Save as draft
4. Remove a phone number from a VAMC facility page
   - [x] Edit the page you just created
   - [x] In the "Mental health phone number" section click the Remove button
   - [x] Click confirm removal button (this is technically optional)
   - [x] Add a revision message and save as draft
5. Add a phone number 
   - [x] Edit the page you just created
   - [x] In the "Mental health phone number" section click the Add Phone number button
   - [x] Confirm that the correct phone number fields populate

### Check  VAMC Facility - Mental Health Phone Number TOGGLE OFF
1. [Log in](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/) as an admin
2. Go to [Feature Toggle page](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/admin/config/system/feature_toggle) 
   - [ ] Uncheck FEATURE_TELEPHONE_MIGRATION_V1
   - [ ] Click Save
3. Go to Content > Add Content > [VAMC Facility](https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/node/add/health_care_local_facility)
4. Notice the slightly changed interface compared to prod

### Check other phone number paragraph types to make sure we didn't break their styling
- VAMC System VA Police
- VBA Facility Service (service location)


### Definition of Done

- [x] Automated tests have passed.
- [x] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

jilladams commented:
# QA notes

## Basic behavior
- Content with no phone number: add phone, save:  🟢
- Content with a phone number: remove phone number, save: 🟢 

## Label field weirdness 🔴 
On Staff profile, I removed an existing phone number, added a new one and saved with max char length to test. After removing phone, clicking to Add phone, I see a Label field (across all content types). To reproduce: 

* Remove an existing phone # 
* Add phone - new fields appear: Type, Phone Number, Extension number, Label. 
  * Refresh page without populating fields: 
    * Label field disappears
  * OR, save content in the field. 
    * Node view shows the Label field value
    * Node edit hides the Label field

https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/node/73936
![Screenshot 2024-10-25 at 3 31 43 PM](https://github.com/user-attachments/assets/186a86e9-820f-43d7-b0a1-04ec0a8f66c9)

https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/node/73936/edit
![Screenshot 2024-10-25 at 3 31 37 PM](https://github.com/user-attachments/assets/2cf860f7-163a-4d85-a234-6f5be7fc1e8a)

Test content (FE is building on Fri pm and can be verified on Monday):
* Staff profile node: https://web-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/staff-profiles/esther-reece
  * Staff profile FE: https://web-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/staff-profiles/esther-reece
* B&I node: https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/central-western-massachusetts-health-care/billing-and-insurance
  * B&I FE: https://web-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/central-western-massachusetts-health-care/billing-and-insurance
* VA clinic MH: https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/spokane-health-care/locations/spokane-valley-va-clinic
  * FE: https://web-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/spokane-health-care/locations/spokane-valley-va-clinic
* Mobile clinic: https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/southern-arizona-health-care/locations/tucson-va-mobile-clinic
  * https://web-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/southern-arizona-health-care/locations/tucson-va-mobile-clinic
* VAMC: https://pr19607-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/alaska-health-care/locations/joint-base-elmendorf-richardson-va-medical-center
  * https://web-h5ubpjxulma3zyqeq8hayikh80scnqhu.ci.cms.va.gov/alaska-health-care/locations/joint-base-elmendorf-richardson-va-medical-center

This probably blocks the PR, until we prove whether the label prints in the FE or not.

## Remove confirmation / truncated text 🟡 
Removal confirmation is technically optional per the QA steps. 

When I use the "remove" button to remove, I get a confirmation prompt with truncated text. Loom that shows the experience: https://www.loom.com/share/7f203fbc6650489288d51f85a0e3dd09

![Screenshot 2024-10-25 at 3 15 32 PM](https://github.com/user-attachments/assets/49bb0ff1-3305-4629-8452-4a0f93fe45dd)

This doesn't have to block the PR, but is not great UX debt to add while trying to improve the experience, so I do think it needs follow up. If the confirmation is optional, can we remove it? Or do we need to keep it and address the truncated text? (Updating `.js .paragraph-type-top .paragraph-type-title` to > 25% displays the full text.) 

## Empty Cashier's office accordion on B&I
Observed that Cashier's office accordion is empty. That's a documented defect: #16150

jilladams commented:
I'm a goofy person, and we cannot test what happens in the fe, because the FE isn't using the new phone number structure yet. I think. Not sure what that means for troubleshooting / fixing labels and whether or not this can merge.

davidmpickett commented:
> ## Label field weirdness 🔴
> On Staff profile, I removed an existing phone number, added a new one and saved with max char length to test. After removing phone, clicking to Add phone, I see a Label field (across all content types). To reproduce:
> 
> * Remove an existing phone #
> * Add phone - new fields appear: Type, Phone Number, Extension number, Label.
>   
>   * Refresh page without populating fields:
>     
>     * Label field disappears
>   * OR, save content in the field.
>     
>     * Node view shows the Label field value
>     * Node edit hides the Label field
> 

I also noticed this behavior when I was doing QA. However, this behavior is not specific to my PR. I flipped the feature toggle on [my demo tugboat](https://cms-sivbehcvqkui7ce9b67dwanfbohmenin.demo.cms.va.gov/) and confirmed that the label field shows up if you add, remove, and re-add.

![Screenshot 2024-10-28 at 10 02 58 AM](https://github.com/user-attachments/assets/7b643d18-3228-4092-9869-db8a789ed6b6)

Fixing this should be a separate ticket.

davidmpickett commented:
> ## Remove confirmation / truncated text 🟡
> Removal confirmation is technically optional per the QA steps.
> 
> When I use the "remove" button to remove, I get a confirmation prompt with truncated text. Loom that shows the experience: https://www.loom.com/share/7f203fbc6650489288d51f85a0e3dd09
> 
> ![Screenshot 2024-10-25 at 3 15 32 PM](https://private-user-images.githubusercontent.com/85581471/380337200-49bb0ff1-3305-4629-8452-4a0f93fe45dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMjk4MDksIm5iZiI6MTczMDEyOTUwOSwicGF0aCI6Ii84NTU4MTQ3MS8zODAzMzcyMDAtNDliYjBmZjEtMzMwNS00NjI5LTg0NTItNGEwZjkzZmU0NWRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDE1MzE0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxZWIzYjZlMjc5ODU0OTAxNmIxNzU2NTgwMjMwZGI5ZmIyZGRiMTIzMWNhMTgyODE4YjBhNTgyNDc4YjM5ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rH9eSDF00xFX-2VlLFh1906SqsCOyJmbjcBiIc78XPY)
> 
> This doesn't have to block the PR, but is not great UX debt to add while trying to improve the experience, so I do think it needs follow up. If the confirmation is optional, can we remove it? Or do we need to keep it and address the truncated text? (Updating `.js .paragraph-type-top .paragraph-type-title` to > 25% displays the full text.)
> 

I just committed a change to increase that to 40%, but then that makes this PR need CMS platform review since that's changing a more fundamental property in the system. I'm also not sure how we would go about verifying that increasing this percentage doesn't create problems elsewhere in the CMS.

jilladams commented:
I think this is now ready for code review, @omahane

davidmpickett commented:
> ## Empty Cashier's office accordion on B&I
> Observed that Cashier's office accordion is empty. That's a documented defect: #16150

This change was pretty easy to pull in to my PR. 
- I removed the empty accordion
- I updated the top of this PR to reflect it also resolves 16150
- I updated the QA steps to reflect noticing that fix

jv-agile6 commented on config/sync/core.entity_form_display.node.person_profile.default.yml:
```diff
@@ -258,22 +258,16 @@ content:
     third_party_settings: {  }
   field_telephone:
-    type: paragraphs
+    type: entity_reference_paragraphs
```
Added entity reference paragraph

jv-agile6 commented on config/sync/core.entity_form_display.node.vamc_system_billing_insurance.default.yml:
```diff
@@ -248,22 +248,16 @@ content:
     third_party_settings: {  }
   field_telephone:
-    type: paragraphs
+    type: entity_reference_paragraphs
```
added entity reference paragraph

jv-agile6 reviewed (commented):
Reviewed code configuration in core.entity_form_display.node.person_profile.default.yml and config/sync/core.entity_form_display.node.vamc_system_billing_insurance.default.yml for form and logic. Followed additions and subtractions to code and found no defects.

jv-agile6 reviewed (dismissed):
Code review completed. Approval granted.

laflannery commented:
1. Before a phone number is added, `Phone number` is correctly styled bold. When I add a phone number, this is no longer bold, which makes it harder to pick out when I want to go find it again and edit it:

**Before adding**
![image](https://github.com/user-attachments/assets/74d41e4b-ffb9-4f7e-b951-7d6a81250f5b)

**After adding**
![image](https://github.com/user-attachments/assets/a3c5b51b-0aba-4e2b-9be9-e89b9568bebe)

2. Italics are not in the DS, this should be updated to plain text:
![image](https://github.com/user-attachments/assets/9899588f-a458-4115-b21a-53acd6722a74)

3. The `Remove` button should be updated to say `Remove Phone number` or have an `aria-describedby="edit-field-telephone-0-top-paragraph-type-title--0PvCeESrLmY"` attribute for AT users at least. Currently, because it's be disassociated from the table, I have no idea what I'm removing so I need a new association.

4. The whole thing is not wrapped in fieldset/legend markup, which is actually because it's typically contained within a larger fieldset, except on the Billing and Insurance content type. Are we able to change this so that all the phone fields are wrapped in a `<fieldset> tag and "Phone number" is the `<legend>`? Simplified example below:
![image](https://github.com/user-attachments/assets/f8058a47-6d3e-433c-8960-e6ec4adb5e1d)

omahane commented:
@laflannery I can't seem to figure out where your screenshot is from that shows the phone number without a fieldset.

# Fieldsets
## Staff profile
![Screenshot 2024-12-04 at 5 27 01 PM](https://github.com/user-attachments/assets/79532ce7-e79f-4482-8154-9dd5fc6c5218)

## VAMC Facility
![Screenshot 2024-12-04 at 5 27 08 PM](https://github.com/user-attachments/assets/cc00f0b1-8941-4a68-9d93-9655084a3b36)

## VAMC System Billing and Insurance
![Screenshot 2024-12-04 at 5 27 12 PM](https://github.com/user-attachments/assets/e56ae52e-9dcc-48cf-8af1-49e817a9f69e)

All of these have a fieldset. I wonder if the Staff Profile one (which is what I think your screenshot is from) was before Dave added the fieldset.

# Remove button
How important is it to change this remove button to include what is being removed? We would need to either custom code this or try to find/implement an upstream patch on a widget that is legacy, as the name implies and is thus not in development any longer.

omahane commented:
@laflannery I have made "Phone number" when there's a phone and have made the italics normal when there's no phone. Please confirm when you have a minute when this is rebuilt.

laflannery commented:
Sorry, one more thing. On VAMC Facility content type. When the paragraph field is present I see "Phone number" but then when I remove it this label changes to "Mental health phone number". Was that intentional? Can it remain "Phone number"?
![Screenshot 2024-12-05 at 3 15 27 PM](https://github.com/user-attachments/assets/0b3df318-5ae0-451f-aef9-18a4461daf43)

omahane commented:
> Sorry, one more thing. On VAMC Facility content type. When the paragraph field is present I see "Phone number" but then when I remove it this label changes to "Mental health phone number". Was that intentional? Can it remain "Phone number"? !

This is a known bug in the old paragraph widget, which we are using here:
https://www.drupal.org/project/paragraphs/issues/3067178

If we want, we could go back to using the newer one, which has its own UX oddities.

laflannery commented:
@omahane Then that's fine, it's not a deal breaker, thanks!

dsasser commented on config/sync/core.entity_form_display.node.vamc_system_billing_insurance.default.yml:
```diff
@@ -121,20 +120,6 @@ third_party_settings:
         description: ''
         required_fields: false
         formatter: closed
-    group_cashier_s_office:
```
Noting this change is expected as it resolves #16150 which is included in this PR.

dsasser reviewed (approved):
Code changes look good!

omahane commented:
I can't save a new VAMC System Billing and Insurance:
![Screenshot 2024-12-10 at 12 30 40 PM](https://github.com/user-attachments/assets/23fb86ae-9bd3-44b2-979c-5c5dec7ebb30)

It's fine. The phone validation worked correctly.

omahane commented:
I didn't see the point in testing with the toggle off, as we've already migrated and have moved over to the new paragraph phone field in production.
